drivers/sensor/ti_hdc2xxx/ti_hdc2xxx.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2021 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_hdc2xxx
+
+#include <kernel.h>
+#include <drivers/i2c.h>
+#include <drivers/sensor.h>
+#include <sys/byteorder.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(TI_HDC2XXX, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Register addresses */
+#define TI_HDC2XXX_REG_TEMP		0x00
+#define TI_HDC2XXX_REG_HUMIDITY		0x02
+#define TI_HDC2XXX_REG_MEAS_CFG		0x0F
+#define TI_HDC2XXX_REG_MANUFACTURER_ID	0xFC
+#define TI_HDC2XXX_REG_DEVICE_ID	0xFE
+
+/* Register values */
+#define TI_HDC2XXX_MANUFACTURER_ID	0x5449
+#define TI_HDC2XXX_DEVICE_ID		0x07D0
+
+struct ti_hdc2xxx_config {
+	const struct device *bus;
+	uint16_t i2c_addr;
+};
+
+struct ti_hdc2xxx_data {
+	int32_t t_sample;
+	int32_t rh_sample;
+};
+
+static int ti_hdc2xxx_sample_fetch(const struct device *dev,
+				   enum sensor_channel chan)
+{
+	const struct ti_hdc2xxx_config *drv_cfg = dev->config;
+	struct ti_hdc2xxx_data *drv_data = dev->data;
+	uint16_t buf[2];
+	int rc;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	/* start conversion of both temperature and humidity with the default accuracy (14 bits) */
+	rc = i2c_reg_write_byte(drv_cfg->bus, drv_cfg->i2c_addr, TI_HDC2XXX_REG_MEAS_CFG, 0x01);
+	if (rc < 0) {
+		LOG_DBG("Failed to write measurement configuration register");
+		return rc;
+	}
+
+	/* for 14-bit conversion temperature needs 660us and humidity 610us */
+	k_msleep(2);
+
+	/* temperature and humidity registers are consecutive, read them in the same burst */
+	rc = i2c_burst_read(drv_cfg->bus, drv_cfg->i2c_addr,
+			    TI_HDC2XXX_REG_TEMP, (uint8_t *)buf, sizeof(buf));
+	if (rc < 0) {
+		LOG_DBG("Failed to read sample data");
+		return rc;
+	}
+
+	drv_data->t_sample = sys_le16_to_cpu(buf[0]);
+	drv_data->rh_sample = sys_le16_to_cpu(buf[1]);
+
+	return 0;
+}
+
+
+static int ti_hdc2xxx_channel_get(const struct device *dev,
+				  enum sensor_channel chan,
+				  struct sensor_value *val)
+{
+	struct ti_hdc2xxx_data *drv_data = dev->data;
+	int32_t tmp;
+
+	/* See datasheet "Register Maps" section for more details on processing sample data. */
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		/* val = -40 + 165 * sample / 2^16 */
+		tmp = drv_data->t_sample * 165U;
+		val->val1 = (tmp >> 16) - 40;
+		/* x * 1000000 / 65536 = x * 15625 / 1024 */
+		val->val2 = ((tmp & 0xFFFF) * 15625U) >> 10;
+		break;
+	case SENSOR_CHAN_HUMIDITY:
+		/* val = 100 * sample / 2^16 */
+		tmp = drv_data->rh_sample * 100U;
+		val->val1 = tmp >> 16;
+		/* x * 1000000 / 65536 = x * 15625 / 1024 */
+		val->val2 = ((tmp & 0xFFFF) * 15625U) >> 10;
```

Comment:
Better define constants (165, 15625, 40, etc). I imagine these are scaling and offset constants defined in the datasheet.

Reply:
165, 100 and 40 are indeed scaling and offset factors from the datasheet. I'll introduce them as constants.
15625 is just 1e6 (the factor between val1 and val2 in Zephyr) divided by 64, with the same change applied to the denominator, to avoid using a int64_t variable.